pkg/cloud/aws/services/ec2/subnets.go

```diff
@@ -175,14 +175,56 @@ func (s *Service) describeVpcSubnets() (v1alpha1.Subnets, error) {
 		return nil, errors.Wrapf(err, "failed to describe subnets in vpc %q", s.scope.VPC().ID)
 	}
 
+	routeTables, err := s.describeVpcRouteTablesBySubnet()
+	if err != nil {
+		return nil, err
+	}
+
+	natGateways, err := s.describeNatGatewaysBySubnet()
+	if err != nil {
+		return nil, err
+	}
+	internetGateways, err := s.describeVpcInternetGateways()
+	if err != nil {
+		return nil, err
+	}
+
 	subnets := make([]*v1alpha1.SubnetSpec, 0, len(out.Subnets))
 	for _, ec2sn := range out.Subnets {
+		rt := routeTables[*ec2sn.SubnetId]
+		var hasPublicRoute bool
+		var routeTableId, natGatewayId *string
+		if rt != nil {
+			routeTableId = rt.RouteTableId
+			for _, route := range rt.Routes {
+				if route == nil {
```

Comment:
This could be `if route == nil || route.GatewayId == nil` and remove the if condition below.

Reply:
Also, could a route in the slice actually be `nil`?

Reply:
It's a pointer, so maybe? The data there is more or less straight from the AWS APIs, so if we got a response that looked something like:

```
...
{
  "Routes": [
    {
      "DestinationCidrBlock": "0.0.0.0/0",
      "GatewayId": "igw-zzzz",
      "Origin": "CreateRoute",
      "State": "active"
    },
    null
  ]
}
...
```

Then I would assume that would show up as a nil pointer. Whether AWS is capable of responding that way, I don't know.

Reply:
I think the go-aws-sdk and aws APIs guarantee to not return nil values in slices, so it should be ok to remove the check.